tools/rewriter/SourceRewriter.cpp

```diff
@@ -1117,31 +1072,120 @@ std::optional<Pkey> pkey_from_commands(std::function<std::optional<std::vector<C
+    std::string error;
+    auto db = CompilationDatabase::loadFromDirectory(llvm::Twine(unique_dir).getSingleStringRef(), error);
+    if(!error.empty()) {
+        llvm::errs() << error;
+    }
+    err = remove_directories(unique_dir);
+    return db;
+}
+
+auto ValidateCcDbPathOrDirectory = CLI::Validator(
+        [](std::string &input) {
+            if (CLI::ExistingDirectory(input) == ""s) {
+                input += "/compile_commands.json";
+                llvm::errs() << "warning: Accepting directory containing compile_commands.json. Please pass the path to the JSON directly.\n";
+                return ""s;
+            } else if (CLI::ExistingFile(input) == ""s) {
+                return ""s;
+            } else {
+                return "Argument should be a path to an existing compiler_commands.json compilation database: "s + input;
+            }
+        }, "COMPILE_COMMANDS database", "compile_commands.json or containing dir");
+
+auto ValidateDirectory = CLI::Validator(
+        [](std::string &input) {
+            auto e = CLI::ExistingDirectory(input);
+            if(e != "") {
+              return e;
+            }
+            llvm::SmallString<PATH_MAX> real_path;
+            auto ec = llvm::sys::fs::real_path(input, real_path);
+            if (ec) {
+              return ec.message();
+            }
+            input = std::string(real_path) + "/";
+            return ""s;
+        }, "Directory", "path of a directory");
+
+const std::map<std::string,int> arch_map {{"x86", (int)Arch::X86}, {"aarch64", (int)Arch::Aarch64}};
+
+std::string CompilationDbPath;
+std::vector<std::string> SourceFiles;
+std::vector<std::string> ExtraArgs;
+
 int main(int argc, const char **argv) {
-#if LLVM_VERSION_MAJOR >= 13
-  auto parser_ptr =
-      CommonOptionsParser::create(argc, argv, SourceRewriterCategory);
-  if (!parser_ptr) {
-    auto error = parser_ptr.takeError();
-    llvm::errs() << error;
-    return 1;
+  CLI::App app {"IA2 Source Rewriter"};
+  //CommonOptionsParser options_parser(argc, argv, SourceRewriterCategory);
+  app.add_option("--arch", Target, "Target architecture (x86 or aarch64), x86 by default")
+      ->option_text("x86|aarch64")
+      ->transform(CLI::Transformer(arch_map));
+  app.add_option("--enable-dav1d_get_picture-post-condition", enable_dav1d_get_picture_post_condition,
+      "enable calling the post condition function hardcoded for dav1d_get_picture")
+    ->group("Dav1d-specific");
+  app.add_option("--output-directory", OutputDirectory, "Root directory for output files")
+    ->option_text("<DIR> (REQUIRED)")
+    ->transform(ValidateDirectory)
+    ->required();
+  app.add_option("--root-directory", RootDirectory, "Root directory for input files")
+    ->option_text("<DIR> (REQUIRED)")
+    ->transform(ValidateDirectory)
+    ->required();
+  app.add_option("--output-prefix", OutputPrefix, "Path prefix for output files")
+    ->option_text("<FILENAME-PREFIX> (REQUIRED)")
+    ->required();
```

Comment:
Would it make sense to have a default value for the output prefix? This is just the name used for the various generated `.ld` and `.objcopy` files, right? Seems like we could default that to "wrapper" or "ia2_" or something so that users don't need to specify the flag unless they want to customize the name.

Reply:
I think so, that's the gist of #482, but for this PR I wanted to minimize semantic changes.